random_person.js

```diff
@@ -1,295 +1,456 @@
+                    data = data || extension.data;
+                    if (!data.names[gender]) {
+                        throw new Error(
+                            "No data for the gender '" + gender + '"'
+                        );
+                    }
+                    var t = 0;
+                    var names = data.names[gender];
+                    var i = math.randomint(find_count(names));
+                    var name = find_at_index(names, i);
+                    return name_info(name, gender);
+                });
 
-    /** Choose n random people with the given gender.
-     */
-    var random_people_with_gender = extension.random_people_with_gender = function(gender,n,data) {
-        data = data || extension.data;
-        if(!data.names[gender]) {
-            throw(new Error("No data for the gender '"+gender+'"'));
-        }
-        return weighted_random_subset(data.names[gender],data.totals[gender], n).map(function(x){return name_info(x,gender)});
-    }
+            function maybe_nonbinary(person, data) {
+                if (
+                    data.totals["neutral"] > 0 &&
+                    Math.random() < extension.PROB_NONBINARY
+                ) {
+                    person = name_info(person.name, "neutral");
+                }
+                return person;
+            }
 
-    /** Choose a random person whose name begins with the given letter.
-     */
-    var random_person_with_initial = extension.random_person_with_initial = function(initial,data) {
-        data = data || extension.data;
-        var names = {};
-        var totals = {};
-        Object.keys(data.names).forEach(function(gender) {
-            var gnames = data.names[gender].filter(function(d) { return d.name.slice(0,initial.length)==initial; });
-            var t = 0;
-            gnames.forEach(function(d) {
-                t += d.count;
+            /* Choose a random name and return it along with gender and pronouns.
+             * Gender is chosen uniformly at random from the options available in the data
+             * Returns a dictionary {gender:, name:, pronouns:}, where `pronouns` is a dictionary with keys 'they', 'their' and 'theirs'.
+             */
+            var random_person = (extension.random_person = function (data) {
+                data = data || extension.data;
+                var gender = weighted_choose(data.totals);
+                return maybe_nonbinary(
+                    random_person_with_gender(gender, data),
+                    data
+                );
             });
-            names[gender] = gnames;
-            totals[gender] = t;
-        });
-        var idata = {names: names, totals: totals};
 
-        return random_person(idata);
-    }
-
-    /** Choose n random people whose names each start with distinct letters.
-     */
-    var random_people_with_different_initials = extension.random_people_with_different_initials = function(n,data) {
```

Comment:
changed quite substantially because it could be done faster with the new representation. see green 331